src/utils_latency_config.c

```diff
@@ -0,0 +1,163 @@
+ * Copyright (C) 2013-2016   Florian octo Forster
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Florian octo Forster <octo at collectd.org>
+ *   Pavel Rochnyack <pavel2000 at ngs.ru>
+ */
+
+#include "collectd.h"
+#include "common.h"
+#include "utils_latency_config.h"
+
+int latency_config_add_percentile (const char *plugin, latency_config_t *cl,
+    oconfig_item_t *ci)
+{
+  if ((ci->values_num != 1) || (ci->values[0].type != OCONFIG_TYPE_NUMBER))
+  {
+    ERROR ("%s plugin: \"%s\" requires exactly one numeric argument.",
+             plugin, ci->key);
+    return (-1);
+  }
+
+  double percent = ci->values[0].value.number;
+  double *tmp;
+
+  if ((percent <= 0.0) || (percent >= 100))
+  {
+    ERROR ("%s plugin: The value for \"%s\" must be between 0 and 100, "
+        "exclusively.", plugin, ci->key);
+    return (ERANGE);
+  }
+
+  tmp = realloc (cl->percentile,
+      sizeof (*cl->percentile) * (cl->percentile_num + 1));
+  if (tmp == NULL)
+  {
+    ERROR ("%s plugin: realloc failed.", plugin);
+    return (ENOMEM);
+  }
+  cl->percentile = tmp;
+  cl->percentile[cl->percentile_num] = percent;
+  cl->percentile_num++;
+
+  return (0);
+} /* int latency_config_add_percentile */
+
+int latency_config_add_rate (const char *plugin, latency_config_t *cl,
+    oconfig_item_t *ci)
+{
+
+  if ((ci->values_num != 2)
+    ||(ci->values[0].type != OCONFIG_TYPE_NUMBER)
+    ||(ci->values[1].type != OCONFIG_TYPE_NUMBER))
+  {
+    ERROR ("%s plugin: \"%s\" requires exactly two numeric arguments.",
+             plugin, ci->key);
+    return (-1);
+  }
+
+  if (ci->values[1].value.number &&
+      ci->values[1].value.number <= ci->values[0].value.number)
+  {
+    ERROR ("%s plugin: MIN must be less than MAX in \"%s\".",
+             plugin, ci->key);
+    return (-1);
+  }
+
+  if (ci->values[0].value.number < 0.001)
+  {
+    ERROR ("%s plugin: MIN must be greater or equal to 0.001 in \"%s\".",
```

Comment:
But this is configuring the limits for the "Rate" option, not adding a measurement via `latency_counter_add()`, so this check is not actually triggered, right? Since a latency measurement must be larger than zero, configuring the lower bound with zero effectively means "no lower bound". So `min=0` behaves exactly like `max=0` in that it does not filter out any values because they are too small (too large in the max case). I find this symmetry in the config semantics very appealing.